android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/QuestionnaireActivity.kt

```diff
@@ -202,19 +202,32 @@ open class QuestionnaireActivity : BaseMultiLanguageActivity(), View.OnClickList
       if (readOnly) {
         finish()
       } else {
-        showConfirmAlert(
-          context = this,
-          message = R.string.questionnaire_alert_submit_message,
-          title = R.string.questionnaire_alert_submit_title,
-          confirmButtonListener = { handleQuestionnaireSubmit() },
-          confirmButtonText = R.string.questionnaire_alert_submit_button_title
-        )
+        showFormSubmissionConfirmAlert()
       }
     } else {
       showToast(getString(R.string.error_saving_form))
     }
   }
 
+  fun showFormSubmissionConfirmAlert() {
+    if (questionnaire.experimental)
+      showConfirmAlert(
+        context = this,
+        message = R.string.questionnaire_alert_test_only_message,
+        title = R.string.questionnaire_alert_test_only_title,
+        confirmButtonListener = { handleQuestionnaireSubmit() },
+        confirmButtonText = R.string.questionnaire_alert_test_only_button_title
+      )
+    else
+      showConfirmAlert(
+        context = this,
+        message = R.string.questionnaire_alert_submit_message,
```

Comment:
can we put these messages into meta data on the Questionnaire? then we could load these messages from the Questionnaire, if they don't exist we use a default, and avoid the branching

Reply:
If we add this to questionnaire each questionnaire would have to explicitly add this OR default (Save data) would be used. Here we want to force message 'Data will NOT be saved' incase questionnaire is experimental/test only. We would still need a check that handles separate messages for normal/test-only flow. Keeping it in string.xml also helps to use language specific messages. We can create a ticket to move these to configs in future when we have dealt with internationalization in configs

Reply:
OK that makes sense b/c of i18n

Reply:
> This experimental is a field in questionnaire https://hl7.org/fhir/questionnaire-definitions.html#Questionnaire.experimental. If we want to change it we would have to introduce separate flow specific to this feature.

Oh sorry, I didn't realize that, very good, this looks great then